File: rust/src/day1/solution.rs
struct Elf {
    calories: u32,
}

pub fn day1(data: Vec<String>) -> u32 {
    let elves = get_elves_with_calories(data);
    elves.into_iter().map(|elf| elf.calories).max().unwrap()
}

pub fn day1_2(data: Vec<String>) -> u32 {
    let elves = get_elves_with_calories(data);
    let mut cals = elves
        .into_iter()
        .map(|elf| elf.calories)
        .collect::<Vec<u32>>();
    cals.sort_unstable();
    cals.reverse();
    cals.into_iter().take(3).sum()
}

fn get_elves_with_calories(data: Vec<String>) -> Vec<Elf> {
    let mut elves = vec![Elf { calories: 0 }];
    let mut counter: u32 = 0;
    for line in data {
        if line == "".to_string() {
            elves.push(Elf { calories: 0 });
            counter += 1
        } else {
            elves[counter as usize].calories += line.parse::<u32>().unwrap();
        }
    }
    elves
}

#[test]
fn test_solution1() {
    assert_eq!(
        day1(vec!["1".to_string(), "".to_string(), "2".to_string()]),
        2
    );
}

#[test]
fn test_solution2() {
    assert_eq!(
        day1_2(vec![
            "1".to_string(),
            "".to_string(),
            "2".to_string(),
            "".to_string(),
            "3".to_string()
        ]),
        6
    );
}

File: python/day1/solution.py
from python.read import parse_lines


class Elf:
    calories: int = 0


def solution1(data: list[str]) -> int:
    elves = get_elves_with_calories(data)
    return max(elf.calories for elf in elves)


def solution2(data: list[str]) -> int:
    elves = get_elves_with_calories(data)
    return get_sum_of_top_3(elves)


def get_sum_of_top_3(elves: list[Elf]) -> int:
    sort_elves = sorted(elves, key=lambda x: x.calories, reverse=True)
    print([elf.calories for elf in sort_elves[:3]])
    return sum(elf.calories for elf in sort_elves[:3])


def get_elves_with_calories(data: list[str]) -> list[Elf]:
    elves = [Elf()]
    for line in data:
        if line == "\n":
            elves.append(Elf())
            continue
        elves[-1].calories += int(line)
    return elves


if __name__ == "__main__":
    input_1 = parse_lines("../../inputs/day1.txt")
    print(f"Maximum calories for one elf is: {solution1(input_1)}")
    print(f"Sum of top three calories per elf is: {solution2(input_1)}")
